src/Workspaces/Core/Portable/Indentation/IIndentationService.cs

```diff
@@ -0,0 +1,69 @@
+    /// 
+    /// This pairing provides flexibility to the implementor to compute the indentation results in
+    /// a variety of ways.  For example, one implementation may wish to express indentation of a 
+    /// newline as being four columns past the start of the first token on a previous line.  Another
+    /// may wish to simply express the indentation as an absolute amount from the start of the 
+    /// current line.  With this tuple, both forms can be expressed, and the implementor does not
+    /// have to convert from one to the other.
+    /// </summary>
+    internal struct IndentationResult
+    {
+        /// <summary>
+        /// The base position in the document that the indent should be relative to.  This position
+        /// can occur on any line (including the current line, or a previous line).
+        /// </summary>
+        public int BasePosition { get; }
+
+        /// <summary>
+        /// The number of columns the indent should be at relative to the BasePosition's column.
+        /// </summary>
+        public int Offset { get; }
+
+        public IndentationResult(int basePosition, int offset) : this()
+        {
+            this.BasePosition = basePosition;
+            this.Offset = offset;
+        }
+    }
+
+    internal interface IIndentationService : ILanguageService
+    {
+        /// <summary>
+        /// Determines the desired indentation of a given line.  May return <see langword="null"/> if the
+        /// <paramref name="document"/> does not want any sort of automatic indentation.  May also return
+        /// <see langword="null"/> if the line in question is not blank and thus indentation should
+        /// be deferred to the formatting command handler to handle.
+        /// </summary>
+        IndentationResult? GetDesiredIndentation(Document document, int lineNumber, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Determines indentation for a blank line (i.e. after hitting enter at the end of a line,
+        /// or after moving to a blank line).
+        /// 
+        /// Specifically, this function operates as if the line specified by <paramref name="lineNumber"/>
+        /// is blank.  The actual contents of the line do not matter.  All indentation information is
+        /// determined from the previous lines in the document.
+        /// 
+        /// This is often useful for features which want to insert new code at a certain
+        /// location, indented to the appropriate amount.  This allows those features to
+        /// figure out that position, without having to care about what might already be
+        /// at that line (or further on in the document).
+        /// 
+        /// This function will always succeed.
+        /// </summary>
+        IndentationResult GetBlankLineIndentation(
+            Document document, int lineNumber, FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken);
```

Comment:
This method is removed in: https://github.com/dotnet/roslyn/pull/34001.  it's here only to help enable tehse individual refactoring steps.